core/src/core/ok_verifier.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 The Grin Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! OKVerifier trait for batch verifying outputs and kernels.
+//! We pass a "caching verifier" into the block validation processing with this.
+
+use lru_cache::LruCache;
+
+use core::hash::{Hash, Hashed};
+use core::{Output, TxKernel};
+use util::secp;
+use util::secp::pedersen::{Commitment, RangeProof};
+use util::LOGGER;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Error {
+	Rangeproof,
+	KernelSignature,
+	Secp(secp::Error),
+}
+
+impl From<secp::Error> for Error {
+	fn from(e: secp::Error) -> Error {
+		Error::Secp(e)
+	}
+}
+
+pub trait OKVerifier {
```

Comment:
I'm not sure I like having this in core and I'm not sure I like it facading validation. Wouldn't a bucketed (for each type, range proof or kernel, etc) hash LRU in `util` be simpler? We'd have a few more ifs but it's very localized and it seems making it explicit would also be better.

Reply:
I'm not sure I follow what you're saying.
I get that this probably shouldn't live in core - that makes sense.
But what do you mean by facading validation?

We need to be using the same instance of the verifier everywhere (to take advantage of the shared cache) so we either passit around or use a global singleton and I was trying to avoid introducing a singleton here.
Is that what you are proposing here? Have a singleton in util and just use it where necessary?

Reply:
Oh I think I understand (after taking a step back) - you are suggesting we keep the validation code in core (where it is currently) and pass cache(s) around directly instead of passing some kind of verifier thing around.
And then use the cache in the validation code.

Reply:
Correct, sorry for not being clearer.